UI/Source/OutlineViews/mac/NestedModelCollectionTreeController.swift

```diff
@@ -0,0 +1,129 @@
+            self?.handleCollectionEvent(event)
+        }
+    }
+
+    internal func isExpandable(_ path: IndexPath) -> Bool {
+        let model = modelAtIndexPath(path)
+        let containingNode = findOrCreateNode(path.dropLast())
+        return containingNode.modelCollection.isModelExpandable(model)
+    }
+
+    internal func countOfChildNodes(_ path: IndexPath) -> Int {
+        return findOrCreateNode(path).modelCollection.models.count
+    }
+
+    internal func modelAtIndexPath(_ path: IndexPath) -> Model {
+        guard !path.isEmpty else { Log.fatal(message: "Empty path passed to modelAtIndexPath()") }
+        let containingNode = findOrCreateNode(path.dropLast())
+        return containingNode.modelCollection.models[path.last!]
+    }
+
+    internal var indexPath: IndexPath {
+        // TODO:(danielh) This is very inefficient, though thankfully isn't called super often, optimize once there's
+        // enough test coverage to ensure correctness.
+        guard let parent = parent, let modelId = modelId else { return IndexPath() }
+        guard let index = parent.modelCollection.models.index(where: { $0.modelId == modelId }) else {
+            Log.fatal(message: "Parent of model collection no longer knows about child")
+        }
+        return parent.indexPath.appending(index)
+    }
+
+    internal weak var parent: NestedModelCollectionTreeController? = nil
+
+    // MARK: Observable
+
+    /// Description of mutations to model collection tree.
+    ///
+    /// IndexPaths are consistent when interpreted in order of removed, added, updated.
+    struct Event {
+        var removed: [IndexPath]
+        var added: [IndexPath]
+        var updated: [IndexPath]
+    }
+
+    public final var proxiedObservable: GenericObservable<Event> { return observers }
+    private final let observers: ObserverList<Event>
+
+    // MARK: Private
+
+    private let modelId: ModelId?
+    private var childrenCache = [ModelId: NestedModelCollectionTreeController]()
+    private let modelCollection: NestedModelCollection
+    private var modelCollectionObserver: Observer?
+    private var diffEngine = DiffEngine()
+
+    private func handleCollectionEvent(_ event: CollectionEvent) {
+        guard case .didChangeState(let state) = event else { return }
+        let indexPath = self.indexPath
+        let changes = diffEngine.update([state.models])
+        guard changes.hasUpdates else { return }
+        let updatedIndexPaths = changes.updatedModelPaths.map {
+            indexPath.appending($0.itemIndex)
+        }
+        let addedIndexPaths = changes.addedModelPaths.map {
+            indexPath.appending($0.itemIndex)
+        }
+        let removedIndexPaths = changes.removedModelPaths.map {
+            indexPath.appending($0.itemIndex)
+        }
+        let modelIds = Set(modelCollection.models.map({ $0.modelId }))
+        childrenCache = childrenCache.filter {
```

Comment:
Short answer: Cause `childrenCache` is a map (`[ModelId:ModelCollection]`) so can't use the indices

Longer answer: My main goal here was to avoid recreating the children model collections as much as possible. Two solutions I considered were: 

1. Keeping an array of ModelCollection? with nil values padding for the ones we haven't expanded yet, and then modifying that list to keep it in sync (same indices) as the model collection when deletes/inserts/moves happen
2. (what I actually did) Keeping a map of `[ModelId: ModelCollection]` and just purging the cache when the model they represent is deleted

Option 1 is pretty nice in that you get an efficient implementation of `indexPath` for free. But option 2 was I think a lot less code to get correct, so I went with that to get an initial version done. Will revisit option 1 with the TODO in `indexPath`

Reply:
mstm!